src/lightgbm/src/main/scala/LightGBMUtils.scala

```diff
@@ -50,25 +109,49 @@ object LightGBMUtils {
     idAsInt
   }
 
-  /** Returns the executors.
+  /** Get number of cores from dummy dataset for 1 executor.
+    * Note: all executors have same number of cores,
+    * and this is more reliable than getting value from conf.
     * @param dataset The dataset containing the current spark session.
-    * @return List of executors in host:port format.
+    * @return The number of cores per executor.
     */
-  def getExecutors(dataset: Dataset[_]): Array[(String, Int)] = {
+  def getNumCoresPerExecutor(dataset: Dataset[_]): Int = {
+    val spark = dataset.sparkSession
+    import spark.implicits._
+    spark.range(0, 1).map(_ => java.lang.Runtime.getRuntime.availableProcessors).collect.head
+  }
+
+  private def getDriverHost(dataset: Dataset[_]): String = {
+    val blockManager = BlockManagerUtils.getBlockManager(dataset)
+    blockManager.master.getMemoryStatus.toList.flatMap({ case (blockManagerId, _) =>
+      if (blockManagerId.executorId == "driver") Some(blockManagerId.host)
+      else None
+    }).head
+  }
+
+  /** Returns a list of executor id and host.
+    * @param dataset The dataset containing the current spark session.
+    * @param numCoresPerExec The number of cores per executor.
+    * @return List of executors as an array of (id,host).
+    */
+  def getExecutors(dataset: Dataset[_], numCoresPerExec: Int): Array[(Int, String)] = {
     val blockManager = BlockManagerUtils.getBlockManager(dataset)
     blockManager.master.getMemoryStatus.toList.flatMap({ case (blockManagerId, _) =>
       if (blockManagerId.executorId == "driver") None
-      else Some((blockManagerId.host, blockManagerId.executorId.toInt))
+      else Some((blockManagerId.executorId.toInt, blockManagerId.host))
     }).toArray
   }
 
-  /** Returns the number of executors.
+  /** Returns the number of executors * number of cores.
     * @param dataset The dataset containing the current spark session.
-    * @return The number of executors.
+    * @param numCoresPerExec The number of cores per executor.
+    * @return The number of executors * number of cores.
     */
-  def getNumExecutors(dataset: Dataset[_]): Int = {
-    val executors = getExecutors(dataset)
-    if (!executors.isEmpty) executors.length
+  def getNumExecutorCores(dataset: Dataset[_], numCoresPerExec: Int): Int = {
+    val executors = getExecutors(dataset, numCoresPerExec)
+    if (!executors.isEmpty) {
+      executors.length * numCoresPerExec
+    }
```

Comment:
Nit: put closing brace and else on same line

Reply:
done

Reply:
wowee you're fast! :)